Add or remove components to libraries from HopsanGUI

robbr48 opened the pull request:
Changes/features:
- Library version 0.3: Only loads components with CAF files specified in library XML
- Libraries always show up in GUI even if components failed to load in HopsanCore
- Non-loaded components have red font and cannot be added to models
- Create new libraries
- Add new components using wizard dialog
- Add existing components (CAF files must be inside library folder)
- Remove component to libraries, with option to also delete files
- Removed all dual-view code (deprecated)
- Minor fixes to library widget behaviour

robbr48 commented:
@peterNordin Do you have time to review this? Especially the new library version could be important to discuss.

peterNordin commented:
@robbr48 yes I will have a look over the weekend

peterNordin commented on HopsanGUI/LibraryHandler.cpp:
```diff
@@ -728,46 +1385,42 @@ bool LibraryHandler::loadLibrary(SharedComponentLibraryPtrT pLibrary, LibraryTyp
                     // Success, add component to library
-                    if (success)
-                    {
+                    if (true) {
```
could remove this if statement since true

peterNordin commented on HopsanGUI/LibraryHandler.h:
```diff
@@ -54,6 +58,7 @@ class ComponentLibrary
 {
 public:
     QString id;
+    QString version;
```
later we should try to reuse the library and component types from the generator library, but some linking changes are needed

robbr48 replied:
I agree. There are more specification classes that are duplicated in HopsanGUI, HopsanCore and HopsanGenerator.

peterNordin commented on HopsanGUI/LibraryHandler.h:
```diff
@@ -84,6 +113,46 @@ class ComponentLibraryEntry
     SharedComponentLibraryPtrT pLibrary;        //!< What library the object belongs to
     HiddenVisibleEnumT visibility=Hidden;       //!< Visible or hidden in library widget
     QStringList displayPath;                    //!< The display path in the library widget
+    EnabledDisabledEnumT disabled=Enabled;      //!< Is component enabled or disabled?
```
This variable name is confusing

robbr48 replied:
It did make sense when it was a Boolean, but I changed to enum to make it less confusing :man_facepalming:

peterNordin commented on HopsanGUI/Widgets/LibraryWidget.cpp:
```diff
@@ -650,231 +515,199 @@ void LibraryWidget::handleItemClick(QTreeWidgetItem *item, int column)
+            QDesktopServices::openUrl(QUrl("file:///" + path));
+        }
+        else if(pReply == pEditXMLAction) {
+            if(!isComponentItem(item)) {
```
I would like to have a isComponentLibraryItem help function here, for clarity

robbr48 replied:
How do you mean? Just rename the `isComponentItem`function?

peterNordin replied:
No, in this case it checks if the item is not a component then it must be a library. It would be easier to read if the code said 'if isLibrary'.
But I realised now that this also works on sub directories. So maybe not an.issue after all. But its a bit strange to be able to edit XML for a sub-directory, but not such a big deal

peterNordin commented on HopsanGUI/Widgets/LibraryWidget.cpp:
```diff
@@ -650,231 +515,199 @@ void LibraryWidget::handleItemClick(QTreeWidgetItem *item, int column)
+        else if(pReply == pEditCodeAction) {
+            if(!isComponentItem(item)) {
+                //Edit library source files
+                SharedComponentLibraryPtrT pLibrary = mItemToLibraryMap[item];
```
not sure if this one should be editable (maybe read-only), since it should be auto generated

robbr48 replied:
I will make it const then for clarity.

peterNordin replied:
I meant, in HoLC I think it showed a meaasage in the editor that you should not edit the file since it will be overwriten everytime you change anything

robbr48 replied:
That's true. We don't have the read-only feature in HopsanGUI text editor yet. You may also want to edit this file, for example for temporarily commenting out a component that makes compilation fail. 

I guess we should not overwrite the entire file when we add/remove components, but instead append/remove the specific lines of code. This requires some clever code parsing though.

peterNordin commented on HopsanGenerator/include/hopsangenerator.h:
```diff
@@ -28,6 +28,12 @@ extern "C" {
 
     HOPSANGENERATOR_DLLAPI bool callExeExportGenerator(const char* outputPath, void* pHopsanSystem,  const char* const externalLibraries[], const int numLibraries, const char* hopsanInstallPath, const char* compilerPath, int architecture=64, messagehandler_t messageHandler=0, void* pMessageObject=0);
+
+    HOPSANGENERATOR_DLLAPI bool callAddComponentToLibrary(const char* libraryXMLPath, const char* typeName, const char* displayName, const char* cqsType, const char * const constantNames[], const int numConstantNames, const char * const constantDisplayNames[], const int numConstantDisplayNames, const char * const constantUnits[], const int numConstantUnits, const char * const constantInits[], con ...
```
nice function :) actually we should invert this relationship so that the GUI uses the Qt api directly instead of going through the c-api. The C-api is only required when used from HopsanCLI. Then we could reduce the C-api to only include functions for those things needed by the CLI. I will do that at some point. (no change needed now)

robbr48 replied:
Sounds good to me!

peterNordin commented on HopsanGenerator/src/HopsanGeneratorLib.cpp:
```diff
@@ -336,3 +337,248 @@ bool callExeExportGenerator(const char* outputPath, void* pHopsanSystem, const c
+
+    //Generate source file for new component
+    ComponentSpecification compSpec;
+    compSpec.typeName = typeName;
```
not sure if needed, but subtype can no be specified here, so only "main" components can be added. Later typename should be capable of containing full type with some separator. Ex: "HydraulicOrific:MySpecialOrifice", but for now subtype is needed. But maybe not int this case because you can not add such component via dialog anyway I assume.

peterNordin commented on HopsanGenerator/src/generators/HopsanGeneratorBase.cpp:
```diff
@@ -930,6 +932,73 @@ bool HopsanGeneratorBase::generateCafFile(QString &rPath, ComponentAppearanceSpe
+}
+
+bool HopsanGeneratorBase::generateLibrarySourceFile(const ComponentLibrary &lib)
+{
```
I think that this function duplicates some existing functionality inside the ComponentLibrary utility class. Check the "generateRegistrationCode" function

robbr48 replied:
Yes it does, but they have only two lines in common, and the utility function adds annoying comments about which library the component belongs to. These comments makes sense if we include a component in e.g. an FMU, but not in the source file for the actual library.

peterNordin replied:
The purpose of the comments was to make it easier for anyone who wants to edit the library manually after it has been created. But sure the number of comment can likely be reduced.
We can merge those two later, But I prefer to have one generator instead of two for the same thing.

peterNordin commented on hopsangeneratorgui/include/hopsangeneratorgui/hopsangeneratorgui.h:
```diff
@@ -49,6 +49,16 @@ class HopsanGeneratorGUI : public QWidget
 
+    bool addComponentToLibrary(const QString &libraryXmlPath, const QString &cafPath);
+
+    bool addComponentToLibrary(const QString &libraryXmlPath, const QString &typeName, const QString &displayName, const QString &cqsType,
```
We should use a class with all of these as members instead. But right now we have one version in the generator and another in the GUI. So we can wait until they are unified.

peterNordin commented on hopsangeneratorgui/src/hopsangeneratorgui.cpp:
```diff
@@ -525,6 +530,91 @@ bool HopsanGeneratorGUI::checkComponentLibrary(const QString& libraryXMLPath)
+    const auto libpath = libraryXmlPath.toStdString();
+    MessageForwarder forwarder(lw->widget());
+    const auto xmlpath = libraryXmlPath.toStdString();
+    const auto typenamestr = typeName.toStdString();
```
subtype is missing, but I am not sure if it is needed in this case anyway, when you have written a new component. Also is should be merged into typename.... later

robbr48 replied:
You can easily work around this by first creating the component and then modifying it's CAF file to add the subtype. We can perhaps add this to the generator later.

peterNordin commented:
I did some tests:

When loading the ExtensionLibrary I get the following warning (it conflicts with it self)
A component with type name 'Subsystem|Initial_Step' was already registered by library 'HopsanExtensionLibrary'. Ignoring version in library 'HopsanExtensionLibrary'.

Choosing the following for a library opens the dir/file for the first subcomponent instead of the library:
Open Containing Folder
Edit component appearance
Edit source code

I am unable to find how to create a new library or to insert or remove components, those context menu options never appear for me.
I tried when loading the ExtensionLibrary

It is possible to chose  on sub-directory items:
Edit component appearance
Edit source code

The red fading in the color of the terminal / messages does not work for me when a componetn library fails to compile

When a ComponentLibvrary fails to compile it is not added to the list, I think that according to the new code, the should be added anyway
I tried introducing an error into the ExampleComponetnLibrary

robbr48 commented:
@peterNordin 

> When loading the ExtensionLibrary I get the following warning (it conflicts with it self)
> A component with type name 'Subsystem|Initial_Step' was already registered by library 'HopsanExtensionLibrary'. Ignoring version in library 'HopsanExtensionLibrary'.

I will investigate. As you noticed above, I did not yet consider subtypes. This is probably related.

> Choosing the following for a library opens the dir/file for the first subcomponent instead of the library:
> Open Containing Folder
> Edit component appearance
> Edit source code

Are you sure? For me the correct XML and source files are opened. The containing folder is wrong for the top-level library folder though. 
 
> I am unable to find how to create a new library or to insert or remove components, those context menu options never appear for me.
> I tried when loading the ExtensionLibrary
>
> It is possible to chose on sub-directory items:
> Edit component appearance
> Edit source code

Strange. When I load the ExtensionLibrary I do get the "Add New Component" and "Add Existing Component" entries if a right-click on a library folder (the ones with green icons).

 
> The red fading in the color of the terminal / messages does not work for me when a componetn library fails to compile

Is this only a cosmetic problem? For me it seems to work.
 
> When a ComponentLibvrary fails to compile it is not added to the list, I think that according to the new code, the should be added anyway
> I tried introducing an error into the ExampleComponetnLibrary

Again, this works for me. I tried introducing errors in both main source file and a component source file. The library is loaded, and all components are red.

robbr48 commented on HopsanGUI/LibraryHandler.cpp:
```diff
@@ -351,6 +374,603 @@ const QVector<SharedComponentLibraryPtrT> LibraryHandler::getLibraries(const QSt
+    }
+
+    //Validate type name
+    ComponentSpecification spec = getSpecification();
```
Yes, there is an `isNameValid()` function, but it does not check that the first character is a letter. We can still use it though, and check the first character separately.

peterNordin replied:
Maybe create a isTypenNameValid utility, by the way, why must the first char be a letter?
But I agre it should.

robbr48 replied:
The component name is used as class name, and according to C++ standard, class names must start with either a letter or an underscore. I decided to only allow letters in this cases, since underscore makes no sense at the beginning of a type name.

peterNordin replied:
OK makes sence

robbr48 commented:
@peterNordin The problem with the extenationLibrary is actually a problem with the library itself. It contains two CAF files with the same typename/subtypename. So the error is correct.

peterNordin commented:
Hmm OK I will look in that. Also I will try a clean rebuild and see if I can get stuff working

peterNordin commented:
@robbr48 OK I did a clean rebuild and now all the options appear and fading color works. 

However, I ran into some new problems. When using the "example component library" and adding a new component under the sub-dir hydraulics, the generated library would not compile. The new component hpp include was added without the relative hydarulics/file.hpp path. 

Also the generated library includes debug output (I assume) to std:err that library is being registered. Also no indentation on the actual registration lines (not such a big deal, but anyway...)

Also I got the add new component dialog to crash the program, just adding some random constants, not sure what I did. I remember only filling in the name description and unit (should use quantity dropdown boxes) then clicking something ....

peterNordin commented:
I did not fill in typename and such for the component, just add new constant parameter with name and description, and click "done"

peterNordin commented:
The _savestate model shows up in the recent models list

robbr48 commented:
> The _savestate model shows up in the recent models list

Fixed.

> However, I ran into some new problems. When using the "example component library" and adding a new component under the sub-dir hydraulics, the generated library would not compile. The new component hpp include was added without the relative hydarulics/file.hpp path.

Fixed.

> Also the generated library includes debug output (I assume) to std:err that library is being registered. Also no indentation on the actual registration lines (not such a big deal, but anyway...)

I removed the output, fixed the indentation, and also corrected library name (was hardcoded).

> Also I got the add new component dialog to crash the program, just adding some random constants, not sure what I did. I remember only filling in the name description and unit (should use quantity dropdown boxes) then clicking something ....
> I did not fill in typename and such for the component, just add new constant parameter with name and description, and click "done"

Fixed. It did not check for empty type name before checking first character.
